apps/system/sys_shell.c

```diff
@@ -50,7 +50,7 @@ int main() {
                     INFO("sys_shell: command causes an error");
                 else if (req.argv[req.argc - 1][0] != '&')
                     /* Wait for foreground command to terminate */
-                    grass->sys_recv(NULL, (void*)&reply, sizeof(reply));
+                    grass->sys_wait();
```

Comment:
If I understand correctly, `sys_wait` is only used by the shell and it is waiting for the foreground process to terminate. This seems to be a very narrow scenario and I don't think we should introduce a new system call for it.

Maybe we can do the follwoing:
In `library/servers/servers.h`, do
```c
struct proc_reply {
    enum {
          SPAWN_OK,
          EXIT_OK,
          CMD_ERROR
    } type;
    int pid;
};
```
And then at this point, do
```c
int from, foreground_pid = reply.pid; // Here, reply.pid is the pid of the spawned process
do {
    grass->sys_recv(&from, (void*)&reply, sizeof(reply));
} while (! (from == GPID_PROCESS && reply.pid == foreground_pid) );
```
In other words, whenever a process is killed/exited, `GPID_PROCESS` should send a message to `GPID_SHELL` with the pid of the killed/exited process.

Reply:
The second commit adds these changes.

The third commit updates `sys_recv` so that a process may specify who can send to them, where an argument of `GPID_ALL` may be passed in to receive a request from any sender. 

It seems flexible enough so that you can choose to not use it at all and only use `GPID_ALL` for any receive, but let me know whether you believe it is necessary.